presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -383,26 +390,56 @@ private synchronized Optional<QueryResults> getCachedResult(long token)
     private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uriInfo, String scheme, DataSize targetResultSize, boolean binaryResults)
     {
         QueryResults queryResults = getNextResult(token, uriInfo, scheme, targetResultSize, binaryResults);
-        if (queryResults.getError() == null || !queryResults.getError().isRetriable()) {
-            return queryResults;
+
+        boolean historyBasedOptimizationEnabled = useHistoryBasedPlanStatisticsEnabled(session) && trackHistoryBasedPlanStatisticsEnabled(session);
+        boolean underOneRetryCount = queryManager.getQueryRetryCount(queryId) < 1;
+        boolean underRetryExecutionTime = queryManager.getQueryInfo(queryId).getQueryStats().getExecutionTime().toMillis() <= getQueryRetryMaxExecutionTime(session).toMillis();
+
+        if (queryResults.getError() != null && historyBasedOptimizationEnabled && underOneRetryCount && underRetryExecutionTime && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            CanonicalPlanWithInfo currentQueryTopLevelPlanHash = queryManager.getFullQueryInfo(queryId).getPlanCanonicalInfo().get(0);
+
+            if (originalBeforeRetryQueryId == null && previousQueryTopLevelPlanHash == null) {
+                originalBeforeRetryQueryId = queryId;
+                previousQueryTopLevelPlanHash = currentQueryTopLevelPlanHash;
+            }
         }
+        else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            CanonicalPlanWithInfo currentQueryTopLevelPlanHash = queryManager.getFullQueryInfo(queryId).getPlanCanonicalInfo().get(0);
+            if (previousQueryTopLevelPlanHash.equals(currentQueryTopLevelPlanHash)) {
+                queryManager.failQuery(queryId, new PrestoException(GENERIC_INTERNAL_ERROR, "Since the plan hashes did not change, your retry query will not execute"));
+                return queryResults;
+            }
```

Comment:
In the else branch, should we do some logging on why the retry codepath wasn't chosen ? (if my understanding is right)

Reply:
@agrawaldevesh Do you think this may cause too much logging? Maybe logging only if retry is enabled?